domain/removal/state/model/machine.go

```diff
@@ -603,6 +591,86 @@ func (st *State) removeBasicMachineData(ctx context.Context, tx *sqlair.TX, mUUI
 	return nil
 }
 
+func (st *State) removeNetNode(ctx context.Context, tx *sqlair.TX, netNodeUUID string) error {
+	type node entityUUID
+	netNodeUUIDRec := node{UUID: netNodeUUID}
+
+	// Start by deleting any IP addresses associated with the net node.
+	deleteIPAddresses := `
+DELETE FROM ip_address
+WHERE net_node_uuid = $node.uuid;
+ `
+
+	stmt, err := st.Prepare(deleteIPAddresses, node{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	if err := tx.Query(ctx, stmt, netNodeUUIDRec).Run(); err != nil {
+		return errors.Errorf("deleting net node ip addresses %q: %w", deleteIPAddresses, err)
+	}
+
+	// Get all link-layer devices associated with the net node.
+	selectDevices := `
+SELECT uuid AS &entityUUID.uuid
+FROM   link_layer_device
+WHERE  net_node_uuid = $node.uuid
+`
+
+	devStmt, err := st.Prepare(selectDevices, entityUUID{}, netNodeUUIDRec)
+	if err != nil {
+		return errors.Errorf("preparing devices for deletion statement: %w", err)
+	}
+
+	var devsToDelete []entityUUID
+	if err := tx.Query(ctx, devStmt, netNodeUUIDRec).GetAll(&devsToDelete); err != nil {
+		if !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("running devices for deletion query: %w", err)
+		}
+	}
+
+	llDevicesToDelete := uuids(transform.Slice(devsToDelete, func(d entityUUID) string { return d.UUID }))
+
+	// Delete all relations that reference the link-layer devices and the devices themselves.
+	deleteDevicesRelations := []string{
+		`DELETE FROM link_layer_device_parent WHERE device_uuid IN ($uuids[:]) OR parent_uuid IN ($uuids[:])`,
+		`DELETE FROM provider_link_layer_device WHERE device_uuid IN ($uuids[:])`,
+		`DELETE FROM link_layer_device_dns_domain WHERE device_uuid IN ($uuids[:])`,
+		`DELETE FROM link_layer_device_dns_address WHERE device_uuid IN ($uuids[:])`,
+		`DELETE FROM link_layer_device WHERE uuid IN ($uuids[:])`,
+	}
+	for _, query := range deleteDevicesRelations {
+		stmt, err := st.Prepare(query, llDevicesToDelete)
+		if err != nil {
+			return errors.Capture(err)
+		}
+
+		if err := tx.Query(ctx, stmt, llDevicesToDelete).Run(); err != nil {
+			return errors.Errorf("deleting reference to machine network in query %q: %w", query, err)
+		}
+	}
+
+	// Delete the net node.
+
+	deleteByNetNode := []string{
+		// TODO (stickupkid): We need to ensure that no unit is still using this
+		// net node. If it is, we need to return an error.
+		"DELETE FROM net_node WHERE uuid = $node.uuid",
+	}
+
+	for _, query := range deleteByNetNode {
+		stmt, err := st.Prepare(query, node{})
+		if err != nil {
+			return errors.Capture(err)
+		}
+
+		if err := tx.Query(ctx, stmt, netNodeUUIDRec).Run(); err != nil {
+			return errors.Errorf("deleting reference to machine network in query %q: %w", query, err)
+		}
```

Comment:
Can we check if we receive a FK constraint error, we return `removalerrors.RemovalJobIncomplete` - that should at least allow a more graceful recovery.

Reply:
Added.